### PR TITLE
[Serialization] Load non-public transitive dependencies on `@testable` imports

### PR DESCRIPTION
**Pull request by xymus**
A `@testable` import allows a client to call internal decls which may refer to non-public dependencies. To support such a use case, load non-public transitive dependencies of a module when it's imported `@testable` from the main module.

We apply this behavior only to `@testable` imports from the main module being built, not transitively trough `@testable` imports of a library. It should support the vast majority of `@testable` uses and provide a reasonable behavior without surprises.

This replaces the previous behavior where we loaded those dependencies for any modules built for testing. This would load more modules for any debug build, opening the door to a different behavior between debug and release builds. In contrast, applying this logic to `@testable` clients will only change the behavior of test targets.

rdar://107329303

---

The last commit, `[Serialization] Load non-public transitive dependencies on @testable imports`, is the most relevant to review this PR. The first 2 commits refactor existing code to allow loading dependencies in two phases.

### CODE REVIEW

**Comment by xymus** [comment]
@swift-ci Please smoke test

File: test/Serialization/access-level-import-dependencies.swift [hunk 1]
```diff
@@ -95,4 +95,24 @@ import PrivateDep
...
 // RUN: %target-swift-frontend -typecheck %t/ClientOfNonPublic.swift -I %t \
+// RUN:   -Rmodule-loading 2>&1 | %FileCheck -check-prefix=HIDDEN-DEP %s
+
+/// With testable imports, transitive dependencies are required.
+//--- TestableClientOfPublic.swift
+@testable import PublicDep
+
+//--- TestableClientOfNonPublic.swift
+@testable import PackageDep // expected-error {{missing required module 'HiddenDep'}}
+@testable import InternalDep // expected-error {{missing required module 'HiddenDep'}}
```
Comment (elsh) [inline test/Serialization/access-level-import-dependencies.swift:106]: Might be worth adding a test with `@_exported` import as well?

**Review by elsh — APPROVED** [review]

File: lib/Serialization/ModuleFileSharedCore.h [hunk 2]
```diff
@@ -614,11 +614,17 @@ class ModuleFileSharedCore {
   ///
   /// If \p packageName is set, transitive package dependencies are loaded if
   /// loaded from the same package.
+  ///
+  /// If \p forTestable, get the desired loading behavior for a @testable
+  /// import. So it reports non-public dependencies as required for a testable
+  /// client can access internal details, which in turn can reference those
+  /// non-public dependencies.
```
Comment (bnbarham) [inline lib/Serialization/ModuleFileSharedCore.h:621]: I think the `so` is in the wrong position here, maybe:
```
<-So it> Reports ... as required for a testable client <+so it> can access
```

**Comment by bnbarham** [comment]
Worth adding a WMO test where one file has a @testable import and the other doesn't?

**Comment by xymus** [comment]
@swift-ci Please smoke test

File: test/Serialization/access-level-import-dependencies.swift (hunk 1 above)
Comment (xymus) [reply test/Serialization/access-level-import-dependencies.swift:106]: I've added a test about it. From the commit message:

A scenario where module A `@testable` imports module B, which `@_exported` imports module C. In this case, non-public imports of B are loaded, but not those of C. This is a limitation of the current implementation and could be improved upon in the future.

**Pull request merged**